### PR TITLE
SpMxSpM Single Precision CSR format functionality added

### PR DESCRIPTION
**Pull request by kvaragan**
1.) SpMxSpM code added to the clSPARSE library
2.) Benchmark code of SpMxSpM added
3.) test-blas3.cpp - new TestCases added to test SpMxSpM,
4.) Includes utility functions to fill CSR values with different types of values, for example random powers of 2. Used this utility to test SpMxSpM - perfectly matches with CPU implementation with no precision errors.
Note: Benchmarks & test-blas3 only support Single Precision. Benchmarks only support square matrices.

5.) "ExternalMTX.cmake" is modified to download benchmark matrices for SpMxSpM

### CODE REVIEW

File: src/library/kernels/SpGEMM_computeNnzCt_kernels.cl [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+*      Processors," Journal of Parallel and Distributed Computing, 2015.
+*  2. Weifeng Liu and Brian Vinter, "An Efficient GPU General Sparse
+*      Matrix-Matrix Multiplication for Irregular Data," Parallel and
+*      Distributed Processing Symposium, 2014 IEEE 28th International
+*      (IPDPS '14), pp.370-381, 19-23 May 2014.
+*  for details. >
+* ************************************************************************ */ 
+
+#define GROUPSIZE_256 256
+
```
Comment (jpola) [inline src/library/kernels/SpGEMM_computeNnzCt_kernels.cl]: This looks similar to spmv algorithm where we are traversing CSR matrix. Maybe it might be accelerated using vectorized version. We did that in csr2dense.

File: src/library/kernels/SpGEMM_computeNnzCt_kernels.cl [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+__kernel
+void compute_nnzCt_kernel(
+        __global const int *d_csrRowPtrA,
+        __global const int *d_csrColIndA,
+        __global const int *d_csrRowPtrB,
+        __global int *d_csrRowPtrCt,
+        const int m)
+{
+    __local int s_csrRowPtrA[GROUPSIZE_256+1];
```
Comment (jpola) [inline src/library/kernels/SpGEMM_computeNnzCt_kernels.cl]: s_csrRowPtrA is not used in this kernel

File: src/library/kernels/SpGEMM_copyCt2C_kernels.cl [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+*  1. Weifeng Liu and Brian Vinter, "A Framework for General Sparse 
+*      Matrix-Matrix Multiplication on GPUs and Heterogeneous 
+*      Processors," Journal of Parallel and Distributed Computing, 2015.
+*  2. Weifeng Liu and Brian Vinter, "An Efficient GPU General Sparse
+*      Matrix-Matrix Multiplication for Irregular Data," Parallel and
+*      Distributed Processing Symposium, 2014 IEEE 28th International
+*      (IPDPS '14), pp.370-381, 19-23 May 2014.
+*  for details. >
+* ************************************************************************ */ 
+
```
Comment (jpola) [inline src/library/kernels/SpGEMM_copyCt2C_kernels.cl]: I would suggest to change #define declarations to the style we currently have. This will make our code consistent.

File: src/tests/resources/sparse_matrix_environment.cpp [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+* ************************************************************************ */
+#include "sparse_matrix_environment.h"
+
+CSRSparseEnvironment::sMatrixType CSRSparseEnvironment::ublasSCsr = CSRSparseEnvironment::sMatrixType();
+
+#ifdef TEST_LONG
```
Comment (jpola) [inline src/tests/resources/sparse_matrix_environment.cpp]: What does the TEST_LONG means?

File: src/library/kernels/SpGEMM_computeNnzCt_kernels.cl (hunk 2 above)
Comment (kvaragan) [inline src/library/kernels/SpGEMM_computeNnzCt_kernels.cl]: Agreed, this is not required, I will remove this.

File: src/library/kernels/SpGEMM_computeNnzCt_kernels.cl (hunk 1 above)
Comment (kvaragan) [inline src/library/kernels/SpGEMM_computeNnzCt_kernels.cl]: I will check this.

File: src/library/blas3/clsparse-spm-spm.cpp [hunk 5]
```diff
@@ -0,0 +1,1009 @@
...
+#define WARPSIZE_NV_2HEAP 64
+#define value_type float
+#define index_type int 
+#define MERGEPATH_LOCAL     0
+#define MERGEPATH_LOCAL_L2  1
+#define MERGEPATH_GLOBAL    2
+#define MERGELIST_INITSIZE 256
+#define BHSPARSE_SUCCESS 0
+ 
+#define CHECKMAL(a, b) \
```
Comment (jpola) [inline src/library/blas3/clsparse-spm-spm.cpp]: I would use C++ to allocate memory and handle the error when there is overflow.

File: src/library/blas3/clsparse-spm-spm.cpp [hunk 6]
```diff
@@ -0,0 +1,1009 @@
...
+    clEnqueueFillBuffer(control->queue(), csrRowPtrCt_d, &pattern, sizeof(cl_int), 0, (m + 1)*sizeof(cl_int), 0, NULL, NULL);
+    
+    int *csrRowPtrCt_h = (int *) malloc((m + 1) * sizeof(int));
+    memset(csrRowPtrCt_h, 0, (m + 1) * sizeof(int));
+    
+    // STAGE 1
+    compute_nnzCt(m, csrRowPtrA, csrColIndA, csrRowPtrB, csrColIndB, csrRowPtrCt_d, control);
+    
+    // statistics
+    int *counter = (int *)malloc(NUM_SEGMENTS * sizeof(int));
```
Comment (jpola) [inline src/library/blas3/clsparse-spm-spm.cpp]: Maybe we can use std::array or std::vector for handling this parameters? This will eliminate use of CHECKMAL macro.

File: src/tests/resources/sparse_matrix_environment.cpp (hunk 4 above)
Comment (kvaragan) [inline src/tests/resources/sparse_matrix_environment.cpp]: This was to test matrices with very long rows/columns compared to other dimension. rows >> cols or cols >> rows.

File: src/library/blas3/clsparse-spm-spm.cpp [hunk 7]
```diff
@@ -0,0 +1,1009 @@
...
+#include "internal/kernel-cache.hpp"
+#include "internal/kernel-wrap.hpp"
+#include "internal/clsparse-internal.hpp"
+
+#include <cmath>
+
+#define GROUPSIZE_256 256
+#define TUPLE_QUEUE 6
+#define NUM_SEGMENTS 128
+#define WARPSIZE_NV_2HEAP 64
```
Comment (jpola) [inline src/library/blas3/clsparse-spm-spm.cpp]: WARPSIZE_NV_2HEAP is used once in this code. Can we get rid of it or use it as variable in proper place? I generally don't like to define variables as #define.

File: src/library/kernels/SpGEMM_copyCt2C_kernels.cl (hunk 3 above)
Comment (kvaragan) [inline src/library/kernels/SpGEMM_copyCt2C_kernels.cl]: Point to any references, so that I will follow that style.

File: src/library/blas3/clsparse-spm-spm.cpp (hunk 7 above)
Comment (kvaragan) [inline src/library/blas3/clsparse-spm-spm.cpp]: Very useful comments, the code is actually written by Shuai, I am planning to make some changes to this code, I will incorporate your comments.

File: src/library/blas3/clsparse-spm-spm.cpp [hunk 8]
```diff
@@ -0,0 +1,1009 @@
...
+
+
+#include "include/clSPARSE-private.hpp"
+#include "internal/clsparse-control.hpp"
+#include "internal/kernel-cache.hpp"
+#include "internal/kernel-wrap.hpp"
+#include "internal/clsparse-internal.hpp"
+
+#include <cmath>
+
```
Comment (jpola) [inline src/library/blas3/clsparse-spm-spm.cpp]: Group size I think should be defined when necessary. In the code I saw the GROUPSIZE_256 is used just before kernel call. Maybe it will be better to put it there as a variable instead of this definition. In addition I thing it would be nice it that GROUPSIZE_256 will be passed to the kernel source by defining -DGROUPSIZE_256 which we are doing in many places. Thank sto that the kernel will be more flexible.

File: src/library/blas3/clsparse-spm-spm.cpp [hunk 9]
```diff
@@ -0,0 +1,1009 @@
...
+#define GROUPSIZE_256 256
+#define TUPLE_QUEUE 6
+#define NUM_SEGMENTS 128
+#define WARPSIZE_NV_2HEAP 64
+#define value_type float
```
Comment (jpola) [inline src/library/blas3/clsparse-spm-spm.cpp]: Does the function design is the only reason why you provided only the interface for single precision? I did not see any specific kernel instruction which forbids to use double precision as a values.

File: src/library/blas3/clsparse-spm-spm.cpp [hunk 10]
```diff
@@ -0,0 +1,1009 @@
...
+#define index_type int
```
Comment (jpola) [inline src/library/blas3/clsparse-spm-spm.cpp]: I saw that in the kernels used for matrix matrix product there are functions for like scan, bitonic_sort. Do you see any obstacles when we switch to the 64 bit integers? Are there any bit operations? I did not see any but can you confirm?

File: src/library/kernels/SpGEMM_EM_kernels.cl [hunk 11]
```diff
@@ -0,0 +1,1201 @@
...
+        {
+            // atomicAdd is not needed since duplicate is not existed in each input row
+            d_val[median] -= val_input;
+            break;
+        }
+    }
+    //return start;
+}
+
+inline
```
Comment (jpola) [inline src/library/kernels/SpGEMM_EM_kernels.cl]: This scan functions makes me wonder, they are almost the same as in ESC_bitonic_kernels.cl but used in slightly different way. Once directly by switch statement and once by calling scan function where switch statement is placed but without `case 16:`

If that is true we can greatly reduce the code there by making this code more general and move it to separate header which will be included during the kernel compilation time.

File: src/library/kernels/SpGEMM_copyCt2C_kernels.cl (hunk 3 above)
Comment (jpola) [inline src/library/kernels/SpGEMM_copyCt2C_kernels.cl]: Look at this code:
https://github.com/clMathLibraries/clSPARSE/blob/master/src/library/kernels/csrmv_general.cl#L18

In addition I think that all parameters should by defined during the kernel call like here: 
https://github.com/clMathLibraries/clSPARSE/blob/master/src/library/blas2/csrmv-vector.hpp#L51

File: src/library/blas3/clsparse-spm-spm.cpp (hunk 9 above)
Comment (kvaragan) [inline src/library/blas3/clsparse-spm-spm.cpp]: Yes, all kernels support double precision.

File: src/library/kernels/SpGEMM_EM_kernels.cl (hunk 11 above)
Comment (kvaragan) [inline src/library/kernels/SpGEMM_EM_kernels.cl]: There are three different algorithms(Heap -sort, bitonic ESC and Merge-Path) used based on the predicted value of nnZ of output matrix. There purpose is same, find duplicate column indices and compress them. Therefore you might see similar functionality across the different kernels.

File: src/include/clSPARSE.h [hunk 12]
```diff
@@ -464,6 +464,15 @@ CLSPARSE_EXPORT clsparseStatus
 clsparseDdense2csr(const cldenseMatrix* A, clsparseCsrMatrix* csr,
                    const clsparseControl control);
 
+//SpM x SpM
```
Comment (kknox) [inline src/include/clSPARSE.h]: Make sure to decorate this function with the appropriate doxygen comments now.  From now on, when we introduce new functions into clsparse.h, they should come with the appropriate documentation.

File: src/benchmarks/clsparse-bench/functions/clfunc_xSpMSpM.hpp [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+
+    double bandwidth() // Need to modify this later **********
+    {
+        //  Assuming that accesses to the vector always hit in the cache after the first access
+        //  There are NNZ integers in the cols[ ] array
+        //  You access each integer value in row_delimiters[ ] once.
+        //  There are NNZ float_types in the vals[ ] array
+        //  You read num_cols floats from the vector, afterwards they cache perfectly.
+        //  Finally, you write num_rows floats out to DRAM at the end of the kernel.
+        return (sizeof(cl_int)*(csrMtx.num_nonzeros + csrMtx.num_rows) + sizeof(T) * (csrMtx.num_nonzeros + csrMtx.num_cols + csrMtx.num_rows)) / time_in_ns();
```
Comment (kknox) [inline src/benchmarks/clsparse-bench/functions/clfunc_xSpMSpM.hpp]: Do you have an idea of what the bandwidth or gflops formula should be?

File: src/library/blas3/clsparse-spm-spm.cpp [hunk 14]
```diff
@@ -0,0 +1,1009 @@
+/* ************************************************************************
+* The MIT License (MIT)
+* Copyright 2014-2015 weifengliu
```
Comment (kknox) [inline src/library/blas3/clsparse-spm-spm.cpp]: I think weifeng requested:
Copyright 2014-2015 University of Copenhagen

File: src/library/blas3/clsparse-spm-spm.cpp (hunk 8 above)
Comment (kknox) [inline src/library/blas3/clsparse-spm-spm.cpp]: I concur

File: src/include/clSPARSE.h [hunk 15]
```diff
@@ -464,6 +464,15 @@ CLSPARSE_EXPORT clsparseStatus
...
+ CLSPARSE_EXPORT clsparseStatus
+        clsparseScsrSpGemm(
+        const clsparseCsrMatrix* sparseMatA,
+        const clsparseCsrMatrix* sparseMatB,
+              clsparseCsrMatrix* sparseMatC,
+        const clsparseControl control );
+
+
```
Comment (kknox) [inline src/include/clSPARSE.h]: Related to Jakubs comments below, how much work would it be to add 'D' precision?  Can you do it now, or do you prefer to wait for a separate pull request?

File: src/library/blas3/clsparse-spm-spm.cpp [hunk 16]
```diff
@@ -0,0 +1,1009 @@
...
+#define MERGEPATH_LOCAL_L2  1
+#define MERGEPATH_GLOBAL    2
+#define MERGELIST_INITSIZE 256
+#define BHSPARSE_SUCCESS 0
+ 
+#define CHECKMAL(a, b) \
+    if(a==NULL) \
+    {     \
+        std::cerr << "Malloc Error:" << b << std::endl; \
+        exit(1); \
```
Comment (kknox) [inline src/library/blas3/clsparse-spm-spm.cpp]: Remove any and all occurrences of exit() from the library.  Libraries should never terminate a process; we need to return an error code.

File: src/library/blas3/clsparse-spm-spm.cpp (hunk 6 above)
Comment (kknox) [inline src/library/blas3/clsparse-spm-spm.cpp]: :+1:

File: src/library/blas3/clsparse-spm-spm.cpp [hunk 17]
```diff
@@ -0,0 +1,1009 @@
...
+                                     0);
+
+
+    copy_Ct_to_C_opencl(counter_one, csrValC, csrRowPtrC, csrColIndC, csrValCt, csrRowPtrCt_d, csrColIndCt, queue_one_d, control);
+    
+    matC->num_rows = m;
+    matC->num_cols = n;
+    matC->num_nonzeros  = nnzC;
+
+    free(csrRowPtrC_h);
```
Comment (kknox) [inline src/library/blas3/clsparse-spm-spm.cpp]: All these EXPLICIT free's and clReleaseMemObjects make me nervous.  See if you can refactor with std::vectors and clsparse::vector

File: src/library/kernels/SpGEMM_EM_kernels.cl [hunk 18]
```diff
@@ -0,0 +1,1201 @@
...
+*      Processors," Journal of Parallel and Distributed Computing, 2015.
+*  2. Weifeng Liu and Brian Vinter, "An Efficient GPU General Sparse
+*      Matrix-Matrix Multiplication for Irregular Data," Parallel and
+*      Distributed Processing Symposium, 2014 IEEE 28th International
+*      (IPDPS '14), pp.370-381, 19-23 May 2014.
+*  for details. >
+* ************************************************************************ */ 
+
+#pragma OPENCL EXTENSION cl_khr_byte_addressable_store : enable
+#pragma OPENCL EXTENSION cl_khr_fp64 : enable
```
Comment (kknox) [inline src/library/kernels/SpGEMM_EM_kernels.cl]: We need to run on hardware that does not support double precision.  This needs preprocessor guards, I think without the #error directive.  Look at #141.

File: src/tests/resources/sparse_matrix_environment.h [hunk 19]
```diff
@@ -0,0 +1,313 @@
...
+#include <cfloat>
+
+//#define TEST_LONG
+//#define NORMALIZE
+
+using CLSE = ClSparseEnvironment;
+
+namespace uBLAS = boost::numeric::ublas;
+
+/**
```
Comment (kknox) [inline src/tests/resources/sparse_matrix_environment.h]: Good job with the doxygen comments!  :wink:

File: src/include/clSPARSE.h (hunk 12 above)
Comment (kvaragan) [inline src/include/clSPARSE.h]: sure I will do that, I remember this but somehow slipped from my mind.

File: src/benchmarks/clsparse-bench/functions/clfunc_xSpMSpM.hpp (hunk 13 above)
Comment (kvaragan) [inline src/benchmarks/clsparse-bench/functions/clfunc_xSpMSpM.hpp]: Yes we have an idea it based on nnzCt_full = total number of non-zeros present in output temporary matrix, unfortunately this parameter is a local variable hidden inside the API. Hence we couldn't use this in the current state of the code. After I modify the code I will try to use this in Flops calculation.

But wondering what flops calculation I should be using for cusparse because they use different algorithm.

File: src/library/blas3/clsparse-spm-spm.cpp (hunk 14 above)
Comment (kvaragan) [inline src/library/blas3/clsparse-spm-spm.cpp]: oh!, I will change it

File: src/include/clSPARSE.h (hunk 15 above)
Comment (kvaragan) [inline src/include/clSPARSE.h]: I would prefer to wait for a separate pull request, Most of the work will be in test-blas3.cpp and benchmark functions and may be some effort needed in the SpMxSpM code.

File: src/library/blas3/clsparse-spm-spm.cpp (hunk 16 above)
Comment (kvaragan) [inline src/library/blas3/clsparse-spm-spm.cpp]: OK

File: src/library/blas3/clsparse-spm-spm.cpp (hunk 17 above)
Comment (kvaragan) [inline src/library/blas3/clsparse-spm-spm.cpp]: sure will consider these inputs in my refactoring tasks.

File: src/library/kernels/SpGEMM_EM_kernels.cl (hunk 18 above)
Comment (jlgreathouse) [inline src/library/kernels/SpGEMM_EM_kernels.cl]: The code I've used to properly guard those pragmas in another kernel is:
```
#if defined(DOUBLE) && __OPENCL_VERSION__ < CL_VERSION_1_2
  #ifdef cl_khr_fp64
    #pragma OPENCL EXTENSION cl_khr_fp64 : enable
  #elif defined(cl_amd_fp64)
    #pragma OPENCL EXTENSION cl_amd_fp64 : enable
  #else
    #error "Double precision floating point not supported by OpenCL implementation."
  #endif
#endif
```

The host-side setup code then needs to pass in `-DDOUBLE` when building the kernel for double precision. If the kernel is only single precision, all of the fp64 pragmas are ignored.

This does not fully fix the problem; there still needs to be checks in the host-side code to ensure that the target device supports double precision before launching an fp64 kernel. Even in OpenCL >= 1.2, fp64 support is only an optional core feature.

You'll probably want to do a similar check against `__OPENCL_VERSION__ < CL_VERSION_1_1` for the cl_khr_byte_addressable_store pragma. This was made into a core feature in OpenCL 1.1. We should probably also check for this feature on the host side and cleanly fail when trying to run against a device that only supports OpenCL 1.0 and does not have this feature.

File: src/library/io/mm-reader.cpp [hunk 20]
```diff
@@ -569,7 +569,7 @@ clsparseSCsrMatrixfromFile(clsparseCsrMatrix* csrMatx, const char* filePath, cls
     MatrixMarketReader< cl_float > mm_reader;
     if( mm_reader.MMReadFormat( filePath ) )
         return clsparseInvalidFile;
-
+#if 0
```
Comment (kknox) [inline src/library/io/mm-reader.cpp]: Why did you disable this code?

Comment (jlgreathouse) [inline src/library/io/mm-reader.cpp]: This code was likely disabled to bypass the bug raised in #144. The change should be reverted, since PR #154 solved the underlying problem.

Comment (kvaragan) [inline src/library/io/mm-reader.cpp]: @kknox  Joseph is right, I disabled it to bypass the bug raised in #144. Just revert it back
